src/hc128.rs

```diff
@@ -0,0 +1,294 @@
+        }
+
+        self.cnt = (self.cnt + 1) & 0x3FF;    
+        ret
+    }
+ 
+    fn next(&mut self) -> u8 {
+        if self.output_index == 0 {
+            let step = self.step();
+            write_u32_le(&mut self.output, step);
+        }
+        let ret = self.output[self.output_index];
+        self.output_index = (self.output_index + 1) & 0x3;
+
+        ret
+    }
+}
+
+fn f1(x: u32) -> u32 {
+    let ret : u32 = x.rotate_right(7) ^ x.rotate_right(18) ^ (x >> 3);
+    ret
+}
+
+fn f2(x: u32) -> u32 {
+    let ret : u32 = x.rotate_right(17) ^ x.rotate_right(19) ^ (x >> 10);
+    ret
+}
+
+impl SynchronousStreamCipher for Hc128 {
+    fn process(&mut self, input: &[u8], output: &mut [u8]) {
+        assert!(input.len() == output.len());
+
+        if input.len() <= 4 {
+            // Process data bytewise
+            for (inb, outb) in input.iter().zip(output.iter_mut()) {
+                *outb = *inb ^ self.next();
+            }
+        } else {
+            let mut data_index = 0 as usize;
+            let data_index_end = data_index + input.len();
+
+            /*  Process any unused keystream (self.buffer) 
+             *  remaining from previous operations */
+            while self.output_index > 0 && data_index < data_index_end {
+                output[data_index] = input[data_index] ^ self.next();
+                data_index += 1;
+            }
+
+            /*  Process input data blockwise until depleted, 
+             *  or remaining length less than block size 
+             *  (size of the keystream buffer, self.buffer : 4 bytes) */
+            while data_index + 4 <= data_index_end {
+                let data_index_inc = data_index + 4;
+
+                // Read input as le-u32
+                let input_u32 = read_u32_le(&input[data_index..data_index_inc]);
+                // XOR with keystream u32
+                let xored = input_u32 ^ self.step();
+                // Write output as le-u32
+                write_u32_le(&mut output[data_index..data_index_inc], xored);
+
+                data_index = data_index_inc;
+            }
+
+            /*  Process remaining data, if any 
+             *  (e.g. input length not divisible by 4) */
+            while data_index < data_index_end {
+                output[data_index] = input[data_index] ^ self.next();
+                data_index += 1;
+            }
+        }
+    }
+}
+ 
+impl Encryptor for Hc128 {
+    fn encrypt(&mut self, input: &mut RefReadBuffer, output: &mut RefWriteBuffer, _: bool)
+            -> Result<BufferResult, SymmetricCipherError> {
+        symm_enc_or_dec(self, input, output)
+    }
+}
+ 
+impl Decryptor for Hc128 {
+    fn decrypt(&mut self, input: &mut RefReadBuffer, output: &mut RefWriteBuffer, _: bool)
+            -> Result<BufferResult, SymmetricCipherError> {
+        symm_enc_or_dec(self, input, output)
+    }
+}
+
+ 
+// #[cfg(test)]
```

Comment:
Actually - I have it backwards. The tests aren't commented out - they are being compiled into non-test builds.